validator/server_arb/execution_run.go

```diff
@@ -50,35 +59,140 @@ func (e *executionRun) PrepareRange(start uint64, end uint64) containers.Promise
+	return stopwaiter.LaunchPromiseThread[[]common.Hash](e, func(ctx context.Context) ([]common.Hash, error) {
+		machine, err := e.cache.GetMachineAt(ctx, machineStartIndex)
 		if err != nil {
 			return nil, err
 		}
-		machineStep := machine.GetStepCount()
-		if position != machineStep {
-			machineRunning := machine.IsRunning()
-			if machineRunning || machineStep > position {
-				return nil, fmt.Errorf("machine is in wrong position want: %d, got: %d", position, machine.GetStepCount())
+		log.Debug(fmt.Sprintf("Advanced machine to index %d, beginning hash computation", machineStartIndex))
+		// If the machine is starting at index 0, we always want to start at the "Machine finished" global state status
+		// to align with the state roots that the inbox machine will produce.
+		var stateRoots []common.Hash
+
+		if machineStartIndex == 0 {
+			gs := machine.GetGlobalState()
+			log.Debug(fmt.Sprintf("Start global state for machine index 0: %+v", gs), "fromBatch", fromBatch)
+			hash := crypto.Keccak256Hash([]byte("Machine finished:"), gs.Hash().Bytes())
+			stateRoots = append(stateRoots, hash)
+		} else {
+			// Otherwise, we simply append the machine hash at the specified start index.
+			stateRoots = append(stateRoots, machine.Hash())
+		}
+		startHash := stateRoots[0]
+
+		// If we only want 1 state root, we can return early.
+		if numDesiredLeaves == 1 {
+			return stateRoots, nil
+		}
+
+		logInterval := numDesiredLeaves / 20 // Log every 5% progress
+		if logInterval == 0 {
+			logInterval = 1
+		}
+
+		start := time.Now()
+		for numIterations := uint64(0); numIterations < numDesiredLeaves; numIterations++ {
+			// The absolute opcode position the machine should be in after stepping.
+			position := machineStartIndex + stepSize*(numIterations+1)
+
+			// Advance the machine in step size increments.
+			if err := machine.Step(ctx, stepSize); err != nil {
+				return nil, fmt.Errorf("failed to step machine to position %d: %w", position, err)
+			}
+			if numIterations%logInterval == 0 || numIterations == numDesiredLeaves-1 {
+				progressPercent := (float64(numIterations+1) / float64(numDesiredLeaves)) * 100
+				log.Info(
+					fmt.Sprintf(
+						"Computing subchallenge progress: %.2f%% - %d of %d hashes needed",
+						progressPercent,
+						numIterations+1,
+						numDesiredLeaves,
+					),
+					"fromBatch", fromBatch,
+					"machinePosition", numIterations*stepSize+machineStartIndex,
+					"timeSinceStart", time.Since(start),
+					"stepSize", stepSize,
+					"startHash", startHash,
+					"machineStartIndex", machineStartIndex,
+					"numDesiredLeaves", numDesiredLeaves,
+				)
 			}
 
+			// If the machine reached the finished state, we can break out of the loop and append to
+			// our state roots slice a finished machine hash.
+			machineStep := machine.GetStepCount()
+			if validator.MachineStatus(machine.Status()) == validator.MachineStatusFinished {
```

Comment:
Consider removing

Reply:
Check if not running to include errored state

Reply:
Assumption about future code: machine stepping when finished might not be free in a future upgrade